crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -43,146 +41,292 @@ pub enum BalanceVariant {
     /// Default format: no symbol, no denomination
     Default(Balance),
     /// Denominated format: symbol and denomination are present
-    Denominated(String),
+    Denominated(DenominatedBalance),
 }
 
 #[derive(Debug, Clone)]
 pub struct TokenMetadata {
     /// Number of denomination used for denomination
-    pub denomination: u128,
+    pub denomination: usize,
     /// Token symbol
     pub symbol: String,
 }
 
+#[derive(Debug, Clone)]
+pub struct DenominatedBalance {
+    value: Decimal,
+    unit: UnitPrefix,
+    symbol: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum UnitPrefix {
+    Giga,
+    Mega,
+    Kilo,
+    One,
+    Milli,
+    Micro,
+    Nano,
+}
+
 impl TokenMetadata {
     /// Query [TokenMetadata] through the node's RPC
     pub async fn query(client: &Client) -> Result<Self> {
         let sys_props = client.rpc().system_properties().await?;
 
         let default_decimals = json!(12);
         let default_units = json!("UNIT");
-        let decimals = sys_props
+        let denomination = sys_props
             .get("tokenDecimal")
```

Comment:
```suggestion
            .get("tokenDecimals")
```
See https://github.com/paritytech/substrate/blob/72b9c36d66e351af9d90cbf8c37e2cf90dd59582/client/chain-spec/res/chain_spec.json#L5k